### PR TITLE
Add workspace abstraction

### PR DESCRIPTION
**Pull request by collijk**
### Background
This is the first change on the path to building a self-contained "AICore" in line with our project plan, and focuses on wrangling the Agent workspace into a coherent abstraction.  It gathers workspace path resolution into a class that the Agent instantiates and uses.  

Due to the prior global nature of the existingworkspace, this change necessarily touches a large number of files, but should radically reduce the scope of certain kinds of future PRs.  Importantly, despite the breadth of this PR, this code change is structured to require no behavior on the user side, however (except, of course, for those users hacking deeply into the codebase).

### Changes

All changes here are in service of creating a workspace abstraction and removing the workspace from the global namespace.

- Add a `workspace_directory` argument to the `Agent` used to construct the Agent's workspace abstraction.
- Collect workspace path resolution into an `Agent` helper function.
- Add a (for now) hidden argument to the app cli to allow the specification of a workspace directory
- Collected workspace and logger file generation into the main function of the cli with appropriate config modification.
- Commands are no longer responsible for path resolution, so all usages of `path_in_workspace` have been removed.
- `WORKSPACE_PATH` no longer exists, so commands (and other code) now access it from the config (for now)
- Added `workspace_path` and `file_logger_path` to the config.
- Added a `workspace` package and module and a `Workspace` class to do the path resolution work.
- Made the conditional skip of the image gen tests unconditional.  They were broken and tests that don't run on CI are not useful tests.
- Added test suite for workspace
- Updated remaining test suite to account for the reduction in global state.

**Could be separate, but I uncovered these issues in testing**
- Added `functools.wraps` decorators to the `command` decorator and the `to allow pytest fixtures to work correctly.

### Documentation

All public functions have complete typing and docstrings and any complex logic is accompanied by comments.

### Test Plan
~40 tests submitted to accompany code changes and all current unit and integration tests pass.

### PR Quality Checklist
- [X] My pull request is atomic and focuses on a single change.
- [X] I have thoroughly tested my changes with multiple different prompts.
- [X] I have considered potential risks and mitigations for my changes.
- [X] I have documented my changes clearly and comprehensively.
- [X] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: autogpt/agent/agent.py
Comment (collijk) [inline autogpt/agent/agent.py:236]: We'll want some kind of introspection mechanism for pathlike command args at some point.

File: autogpt/cli.py
Comment (collijk) [inline autogpt/cli.py:177]: The main purpose of the addition here is to pull the creation of the workspace and file logger file from import time and remove them from the global namespace.

File: autogpt/commands/file_operations.py
Comment (collijk) [inline autogpt/commands/file_operations.py:48]: Now done at app start

File: autogpt/commands/file_operations.py
Comment (collijk) [inline autogpt/commands/file_operations.py:225]: Now done in agent path resolution

File: tests/integration/goal_oriented/test_write_file.py
Comment (collijk) [inline tests/integration/goal_oriented/test_write_file.py:31]: The introduction of a pytest fixture that does its own cleanup removes a ton of boilerplate setup/teardown from tests

File: tests/smoke_test.py
Comment (collijk) [inline tests/smoke_test.py]: Looks like this may have been removed in the past couple of days?  Lmk if we want to decorate this as requiring an API key or delete it.

File: tests/test_image_gen.py
Comment (collijk) [inline tests/test_image_gen.py:17]: Could delete these tests or add a backlog item to fix them.

File: tests/unit/test_file_operations.py
Comment (collijk) [inline tests/unit/test_file_operations.py:26]: If someone knows how to make unittest classes use pytest fixtures, let me know.  Otherwise I think we should add a backlog item to refactor things so we're using pytest only.

File: tests/conftest.py
Comment (ntindle) [inline tests/conftest.py:13]: Make this cross platform by os

File: tests/test_image_gen.py
Comment (ntindle) [reply tests/test_image_gen.py:17]: Add the decorator I suppose

Comment (ntindle) [reply tests/test_image_gen.py:17]: Either way use one of the decorators from the test/utils. Either skip in ci or requires api key

File: tests/test_workspace.py
Comment (ntindle) [inline tests/test_workspace.py:7]: Cross platform this up

File: tests/unit/test_file_operations.py
Comment (ntindle) [inline tests/unit/test_file_operations.py:129]: Why tossed?

**Review by ntindle — CHANGES_REQUESTED** [review]

File: tests/conftest.py
Comment (collijk) [reply tests/conftest.py:13]: This will resolve to a windows path if run on windows.  The `unix-y` part is just a relative path, which is unimportant for testing as everything is nested in a tempdir which will come in as an OS appropriate path from the system root.

File: tests/test_workspace.py
Comment (collijk) [reply tests/test_workspace.py:7]: This will resolve to a windows path if run on windows.  The `unix-y` part is just a relative path, which is unimportant for testing as everything is nested in a tempdir which will come in as an OS appropriate path from the system root.

File: tests/unit/test_file_operations.py
Comment (collijk) [reply tests/unit/test_file_operations.py:129]: This is not a unit test of the file operations.  It's a test of whether `path_in_workspace`, which used to be called by all file operations, raised the appropriate error.  Path resolution has been lifted to the `Workspace` class and is now handled before file operations or any other commands are invoked, so this test is moot.

**Review by ntindle — APPROVED** [review]
Need to run locally before merge. Don’t have laptop handy but code LGTM

**Review by richbeales — APPROVED** [review]

**Pull request merged**
